pkg/disk/v1alpha2/disk.go

```diff
@@ -0,0 +1,251 @@
+type Builder struct {
+	Disk *Disk
+}
+
+// ListBuilder is the builder object for DiskList.
+type ListBuilder struct {
+	DiskList *DiskList
+}
+
+// Predicate defines an abstraction to determine conditional checks against the provided spc instance.
+type Predicate func(*Disk) bool
+
+type predicateList []Predicate
+
+// all returns true if all the predicates succeed against the provided disk instance.
+func (l predicateList) all(c *Disk) bool {
+	for _, pred := range l {
+		if !pred(c) {
+			return false
+		}
+	}
+	return true
+}
+
+// HasAnnotation returns true if provided annotation key and value are present in the provided spc instance.
+func HasAnnotation(key, value string) Predicate {
+	return func(c *Disk) bool {
+		val, ok := c.Object.GetAnnotations()[key]
+		if ok {
+			return val == value
+		}
+		return false
+	}
+}
+
+// IsSparse returns true if the spc is of sparse type.
+func IsSparse() Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == string(apis.TypeSparseCPV)
+	}
+}
+
+// IsDiskActive returns true if the disk is active.
+func IsDiskActive() Predicate {
+	return func(d *Disk) bool {
+		return d.Object.Status.State == DiskActiveState
+	}
+}
+
+// IsUsableDisk returns true if this disk can be used for pool provisioning.
+func IsUsableDisk(usedDiskMap map[string]int) Predicate {
+	return func(d *Disk) bool {
+		return usedDiskMap[d.Object.Name] == 0
+	}
+}
+
+// IsUsableNodeDisk returns true if disk of this node can be used for pool provisioning.
+func IsUsableNodeDisk(usedNodeMap map[string]bool) Predicate {
+	return func(d *Disk) bool {
+		return !usedNodeMap[d.GetNodeName()]
+	}
+}
+
+// IsDiskBelongToNode returns true if the disk belongs to the provided node.
+func IsDiskBelongToNode(nodeName string) Predicate {
+	return func(d *Disk) bool {
+		return d.GetNodeName() == nodeName
+	}
+}
+
+// IsType returns true if the disk is of type same as passed argument
+func IsType(diskType string) Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == diskType
+	}
+}
+
+// IsValidPoolTopology returns true if the topology is valid.
+func IsValidPoolTopology(poolType string, diskCount int) bool {
+	return DefaultDiskCount[poolType]%diskCount == 0
+}
+
+// IsDiskType returns true if the spc is of disk type.
+func IsDiskType() Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == string(apis.TypeDiskCPV)
+	}
+}
+
+// GetNodeName returns the node name to which the disk is attached
+func (d *Disk) GetNodeName() string {
+	return d.Object.GetLabels()[string(apis.HostNameCPK)]
+}
+
+// Filter will filter the disk instances if all the predicates succeed against that spc.
+func (l *DiskList) Filter(p ...Predicate) *DiskList {
+	var plist predicateList
+	plist = append(plist, p...)
+	if len(plist) == 0 {
+		return l
+	}
+
+	filtered := NewListBuilder().List()
```

Comment:
FUTURE: Builder can also have a Filter() as a final function.

```go
NewListBuilder().
  WithFilter(pred2).
  WithFilter(pred2).
  Filter()

NewListBuilder()
  .WithFilters(pred1, pred2)
  .Filter()
```

```go
func (l *DiskList) Filter(p ...Predicate) *DiskList {
  return NewListBuilderForObjectList(l).
    WithFilters(p...).
    Filter()
}
```

Reply:
Yes. Will put it when as and when required.